docs/proposals/vip-4.md

```diff
@@ -0,0 +1,248 @@
+regions. This works fine in many cases, but there are some use cases where we would like to have greater control over 
+the steps of the process. This proposal is to add new configs and hooks which can be used both to monitor and control 
+the push job in a finer-grained manner than is possible today. In particular, this proposal focuses on the way that each 
+individual region is handled.
+
+Currently, the sequence of steps happening within a push job is as follows:
+
+![Push Job Steps](../assets/images/push_job_steps.drawio.svg)
+
+A few notes on the above diagram:
+
+* A new schema will be registered only if the data being pushed does not conform to any already known schema, and 
+  the schema auto-registration config is enabled. If that config is disabled, then an unknown schema leads to job 
+  failure.
+
+* The compression job is optional, and whether it runs or not depends on configurations.
+
+* The data push job writes to a store-version pub sub topic in one of the regions (typically the one which is local to
+  where the push job is running), but all regions start ingesting right away, as soon as the data push job begins 
+  writing. For regions that are remote from the topic the data push job is writing to, the leader servers are performing 
+  replication, while in the region which contains that topic, the replication is a no-op.
+
+* The `SERVING` step is also called "version swap". If there are Da Vinci Clients, they will ingest and swap on their
+  own, and the child controller of that region will wait until all DVC instances have swapped (started serving) before
+  enacting the region-level swap, after which the servers will also start serving the new store-version to clients which
+  perform remote queries. That is why the "DVC Read Traffic SERVING" step is a dependency for the "Server Read Traffic
+  SERVING" step.
+
+## Problem Statement 
+
+Pushing to all regions in parallel makes the push faster, but it also means that if the push causes an issue, the impact
+is going to be global (affecting all regions). It would be desirable to have certain checks and balances that reduce the
+blast radius in cases where the content of the push causes issues. Examples of such issues include:
+
+* Data quality issues:
+  * Incomplete data (due to issues in data generation logic, or in upstream datasets).
+  * Change in semantics (e.g. some embeddings trained by a new ML model / weights / params / etc. are incompatible with 
+    that used at inference-time).
+* Schema-related issues:
+  * Some optional fields which used to always be populated get deleted (or get populated with null) and the reading app 
+    fails due to lack of null checking. This kind of app-side bug can happen even though a schema evolution is fully 
+    compatible.
+* Infra issues:
+  * Larger payloads take more resources, resulting in lack of capacity and thus latency degradation.
+  * Certain types of yet-unknown infra bugs are somehow triggered by a data push.
```

Comment:
Good question... I did not mean that, actually... I meant issues that happen at swap time, such as the issue we had recently where the total size of all schemas exceeded the max response size.

For issues that happen due to ingestion bugs, my understanding is that past issues we've had in that category happened when we turned on some new functionality in a store for the first time (such as zstd compression on a hybrid store) and then did a global push to that store. For that kind of scenario, where test an ingestion-related functionality for the first time, I think we should use the currently-supported flavor of targeted region push (which is not going away as part of this work), and this allows us to ingest in just one region. If the targeted region push succeeds, then we can use that new functionality as part of a global push. If we ramp new ingestion-related functionality using this methodology, we should be able to avoid global outages (at least the kinds we have had in the past)...